### PR TITLE
Shadow Root MutationObserver

### PR DESCRIPTION
**Pull request by alexlorenzi**
## Hey, I just made a Pull Request!

Rollback the changes in #25332 as it caused a [regression bug](https://github.com/backstage/backstage/issues/25414) 

The new fix for the issue raised in #24355 adds MutationObserver in useShadowRootElements so we can trigger updates when the elements under the shadow root change. Previously it was waiting for side-effects from the shadow root itself changing to update. However, that element doesn't change once it's created, new elements are appended to it. The MutationObserver listens for these appends and triggers the relevant side effects.

https://github.com/backstage/backstage/assets/671432/d0496605-498d-4d06-8323-ad81ceb4bdfb

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by benjdlambert — APPROVED** [review]
Approving to unblock, but there's some test failures that need addressing here before we can merge this! :pray:

**Review by PeaWarrior — APPROVED** [review]

**Comment by awanlin** [comment]
Thanks @alexlorenzi 🚀

**Pull request merged**
